### PR TITLE
grafana/ui: Enable panel options update from PanelComponent 

### PR DESCRIPTION
**Pull request by dprokop**
When working on Legend I found a need for modifying panel's options from the panel component itself (i.e. series colors, axis alignment).

This PR introduces two changes:
1. `onOptionsChange` is now passed to the panel component to enable above
2.  PanelProps' `options` property is now of `PanelOptions<T, S>` type.  By default panel options now have optional `series?: S` property, that allows series specific options configuration for a panel. Each panel can specify it's own contract of what are the series' options in it's context. In the case of GraphPanel `S` would be an option object for one particular series(sort of series overrides):
```
interface GraphPanelSeriesOptions {
  displayOnRightYAxis?: boolean; 
  color?: string
}
```

If panel allows series specific options configuration, then it's props would extend PanelProps interface in following way :

```interface MyPanelProps extends PanelProps<MyPanelOptions, MyPanelSeriesOptions> {...}```

The motivation for introducing the PanelOptions type with series being it's first class citizen is the fact, that when removing series from a Panel it will be important to cleanup all series overrides specified for a given object. Without having `series` options available on PanelOptions type, QueriesTab component will not be able to know what to do with series overrides made when manipulating legend (and probably other components that could customise series in the future)

### CODE REVIEW

**Comment by torkelo** [comment]
Interesting idea. 

> Without having series options available on PanelOptions type, QueriesTab component will not be able to know what to do with series overrides made when manipulating legend (and probably other components that could customise series in the future)

Not sure I understand this. In what way does QueriesTab care  / need to know about series overrides?

**Comment by dprokop** [comment]
> Not sure I understand this. In what way does QueriesTab care / need to know about series overrides?

@torkelo let me explain that. Let's say you have changed the colour of A-series visualisation. This information about new color would be stored in the introduced `series` options. Then, when you remove A-series query from the panel, we need to be able to remove the series override from panel options. There are two ways of doing that:
1. Either we introduce `onQueryRemove` prop on QueriesTab, and make sure than panel handles that
2. Or introduce the concept above and let the QueriesTab update panel's options using `onOptionsChange` that the tab already has access to. QueriesTab would simply do the cleanup of the particular series options

Of course in the future, if we see there is a need to perform some kind of side effects when the query is removed, then we can introduce the `onQueryRemove` for handling those effects. But my main driver here is that if something is a series specific option, and the series is no longer available in the panel, then this series options should be cleaned automatically.

**Comment by ryantxu** [comment]
Interesting -- so this is making the concept of series overrides a first class concept on *all* panels.  My two thoughts are:
1. interesting... maybe a good idea
2. Is it possible to support this in the plugin panel directly?  if so, when do we decide to move things to the (rather simple) PanelProps that are applied to all panels?

In a related note, in #16563 I propose passing the queries (ideally immutable) along with the data in PanelProps.  

In the angular panels, essentially anything goes.  that is super powerful.

For sure we will need to expose to let legends set the series color etc.
```
onOptionsChange: (options: PanelOptions<T>) => void;  // really a setOptions(options) function :)
```

**Comment by ryantxu** [comment]
Thinking a little bit more about overrides...

what about having an `seriesOverrides` helper in the core PanelModel?  This would be nice because it could also be migrated using the existing migration callbacks.

**Comment by dprokop** [comment]
> so this is making the concept of series overrides a first class concept on all panels.

Yes, but I like to think about it as a transparent first class concept, hence series is an optional property in the PanelOptions type

>Is it possible to support this in the plugin panel directly? if so, when do we decide to move things to the (rather simple) PanelProps that are applied to all panels?

I think this should be possible, and more elegant solution. See #16566 for alternative proposal.

>what about having an seriesOverrides helper in the core PanelModel

not sure i'm following @ryantxu ?

**Comment by dprokop** [comment]
I think I get what you mean @ryantxu when talking about seriesOverrides in PanelModel. From the migration perspective I think it's beneficial, but from the perspective of the panel "options" manipulation it's gonna be sort of painful, since there would be two methods: one for options manipulation and one for series overrides.

They way I understand we use series overrides now is that basically we are manipulating visualization *options*. And that's why conceptually series overrides belong to panel options form me.

**Comment by torkelo** [comment]
> Then, when you remove A-series query from the panel, we need to be able to remove the series override from panel options. There are two ways of doing that:

This is impossible. A query can return any number of series, named anyway. Unless the panel also stores which query the series name originated from there is no way to remove this option (the angular graph panels does not do any series name -> color clean up). And not sure this clean is preferable, you might want to remove a query to add a different one or modify different one so that it returns more series and want that color option to remain.

But that aside I think to have a common interface for per series options is something I have been thinking about before and would be powerful when switching visualizations to have those per series options be preserved.

**Comment by dprokop** [comment]
@torkelo thanks for clarification! The problem I have encountered was, when I removed query, that the "series override" was applied to the new query result. This is how I did it:

1. Add two queries (A & B)
2. Modify A-series colour and/or y-axis settings
3. Remove A-series query
4. Add new query (results in A-series)

After that, the new series had colour and y-axis settings applied from the series that I have just removed. I see that this behaviour is also present in Angular's Graph panel, which seems to be wrong. But is it really...? While colour being preserved is fine, then y-axis setting seems not to be. 

All of this is performed on test datasource/random walk query. I assume this might result in me being biased.

**Comment by torkelo** [comment]
I think its most likely what you want, if the series name is the same you want the color and y axis to be preserved. Have not had any complaints about this at least in the last 5 years :)

**Comment by ryantxu** [comment]
I like #16566 better.  But I think `WithSeriesOptions` is kinda specific to graph panel and should be flushed out a bit more before getting added to grafana-ui.

I like the idea of generally reusable panel options, but it is a bigger task ;) see #16576 and #16571

File: packages/grafana-ui/src/types/panel.ts [hunk 1]
```diff
@@ -5,11 +5,16 @@ import { ScopedVars } from './datasource';
 
 export type InterpolateFunction = (value: string, scopedVars?: ScopedVars, format?: string | Function) => string;
 
-export interface PanelProps<T = any> {
+export type PanelOptions<T extends {}, S> = T & {
```
Comment (torkelo) [inline packages/grafana-ui/src/types/panel.ts]: Can we remove this State addition to the PaneOptions, not sure it's needed in this PR?

Comment (dprokop) [reply packages/grafana-ui/src/types/panel.ts]: It wasn't state but generic for series options :)

**Pull request closed without merging**